homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+            if (os.path.isdir(os.path.join(recording_path, d)) and
+                d.startswith('recording-'))]
+
+        event_dirs = sorted(
+            all_subdirs,
+            key=lambda x: os.path.getctime(
+                os.path.join(recording_path, x)),
+            reverse=True)
+
+        count = 0
+        event_data = {}
+        for event_dir in event_dirs:
+            if count < start:
+                count += 1
+                continue
+            if count >= start + length:
+                break
+            event_data = {}
+            event_data['directory'] = event_dir
+            event_data['name'] = event_dir
+            event_data['fullPath'] = os.path.join(
+                recording_path,
+                event_dir)
+            event_data['thumbUrl'] = ''
+            event_data['images'] = []
+            event_data['time'] = datetime.datetime.fromtimestamp(
+                os.path.getctime(
+                    event_data['fullPath'])).strftime('%Y-%m-%d %H:%M:%S')
+
+            all_subfiles = [
+                f for f in os.listdir(event_data['fullPath'])
+                if (os.path.isfile(os.path.join(event_data['fullPath'], f)) and
+                    f.startswith('recording_image-'))]
+
+            all_subfiles = sorted(
+                all_subfiles,
+                key=lambda x: os.path.getctime(
+                    os.path.join(event_data['fullPath'], x)),
+                reverse=False)
+
+            for image_file in all_subfiles:
+                full_image_path = os.path.join(
+                    event_data['fullPath'],
+                    image_file)
+
+                image_data = {}
+                image_data['fileName'] = image_file
+                image_data['path'] = (
+                    base_path +
+                    os.path.sep +
+                    event_dir +
+                    os.path.sep +
+                    image_file)
+
+                image_data['url'] = (
+                    'api/saved_camera_image/' +
+                    self.entity_id +
+                    '?image_path=' +
+                    image_data['path'])
+
+                image_data['time'] = datetime.datetime.fromtimestamp(
+                    os.path.getctime(
+                        full_image_path)).strftime('%Y-%m-%d %H:%M:%S')
+                event_data['images'].append(image_data)
+
+            if len(event_data['images']) > 0:
+                thumb_index = math.floor(len(event_data['images'])/2)
+                event_data['thumbUrl'] = (
+                    event_data['images'][thumb_index]['url'])
+
+            events_data.append(event_data)
+            count += 1
+
+        return events_data
+
+    def record_stream(self):
+        """ Records individual frames to disk for a period of time """
+        if self.is_recording:
+            return
+        else:
+            self._is_recording = True
+
+        self.update_ha_state(True)
+        self.send_recording_state()
```

Comment:
`send_motion_state` is part of the `update` method which gets called when calling `update_ha_state` with parameter `True`. Is this by design?

This component is quite large and it is difficult for me to see all the moving parts. I am still working on understanding it